ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/PrivSubscribe.java

```diff
@@ -0,0 +1,64 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.api.jsonrpc.websocket.methods;
+
+import org.hyperledger.besu.ethereum.api.jsonrpc.RpcMethod;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.JsonRpcRequestContext;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.privacy.methods.EnclavePublicKeyProvider;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.response.JsonRpcError;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.response.JsonRpcErrorResponse;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.response.JsonRpcResponse;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.response.JsonRpcSuccessResponse;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.results.Quantity;
+import org.hyperledger.besu.ethereum.api.jsonrpc.websocket.subscription.SubscriptionManager;
+import org.hyperledger.besu.ethereum.api.jsonrpc.websocket.subscription.request.InvalidSubscriptionRequestException;
+import org.hyperledger.besu.ethereum.api.jsonrpc.websocket.subscription.request.PrivateSubscribeRequest;
+import org.hyperledger.besu.ethereum.api.jsonrpc.websocket.subscription.request.SubscriptionRequestMapper;
+import org.hyperledger.besu.ethereum.privacy.PrivacyController;
+
+public class PrivSubscribe extends AbstractPrivateSubscriptionMethod {
+
+  public PrivSubscribe(
+      final SubscriptionManager subscriptionManager,
+      final SubscriptionRequestMapper mapper,
+      final PrivacyController privacyController,
+      final EnclavePublicKeyProvider enclavePublicKeyProvider) {
+    super(subscriptionManager, mapper, privacyController, enclavePublicKeyProvider);
+  }
+
+  @Override
+  public String getName() {
+    return RpcMethod.PRIV_SUBSCRIBE.getMethodName();
+  }
+
+  @Override
+  public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+    try {
+      final PrivateSubscribeRequest subscribeRequest =
+          getMapper().mapPrivateSubscribeRequest(requestContext);
+
+      checkIfPrivacyGroupMatchesAuthenticatedEnclaveKey(
+          requestContext, subscribeRequest.getPrivacyGroupId());
+
+      final Long subscriptionId = subscriptionManager().subscribe(subscribeRequest);
+
+      return new JsonRpcSuccessResponse(
+          requestContext.getRequest().getId(), Quantity.create(subscriptionId));
+    } catch (final InvalidSubscriptionRequestException isEx) {
+      return new JsonRpcErrorResponse(
+          requestContext.getRequest().getId(), JsonRpcError.INVALID_REQUEST);
```

Comment:
Should we also have a branch for an invalid params json rpc error?

Reply:
Yeah, I thought it was being done in the same way we do for the JsonRpc API (https://github.com/hyperledger/besu/blob/master/ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpService.java#L590-L592).

Instead of adding it to each WebSocket method, I think we should add this to the `WebSocketRequestHandler#handle`. So it applies for any WebSocket request.

Reply:
Done! 👍